micro_dl/utils/masks.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+from scipy.ndimage import binary_fill_holes
+from skimage.filters import threshold_otsu
+from skimage.morphology import disk, ball, binary_opening, binary_erosion
+
+
+def create_mask(input_image, str_elem_size=3):
+    """Create a binary mask using morphological operations
+
+    Opening removes small objects in the foreground.
+
+    :param np.array input_image: generate masks from this image
+    :param int str_elem_size: size of the structuring element. typically 3, 5
+    :return: mask of input_image, np.array
+    """
+
+    if np.min(input_image) == np.max(input_image):
+        thr = np.unique(input_image)
+    else:
+        thr = threshold_otsu(input_image, nbins=512)
+    if len(input_image.shape) == 2:
+        str_elem = disk(str_elem_size)
+    else:
+        str_elem = ball(str_elem_size)
+    # remove small objects in mask
+    thr_image = binary_opening(input_image > thr, str_elem)
+    mask = binary_fill_holes(thr_image)
+    return mask
+
+
+def get_unimodal_threshold(input_image):
+    """Determines optimal unimodal threshold
+
+    https://users.cs.cf.ac.uk/Paul.Rosin/resources/papers/unimodal2.pdf
+    https://www.mathworks.com/matlabcentral/fileexchange/45443-rosin-thresholding
+
+    :param np.array input_image: generate mask for this image
+    :return float best_threshold: optimal lower threshold for the foreground
+     hist
+    """
+
+    hist_counts, bin_edges = np.histogram(
+        input_image,
+        bins=256,
+        range=(input_image.min(), np.percentile(input_image, 99.5))
+    )
+    bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2
+
+    # assuming that background has the max count
+    max_idx = np.argmax(hist_counts)
+    int_with_max_count = bin_centers[max_idx]
+    p1 = [int_with_max_count, hist_counts[max_idx]]
+
+    # find last non-empty bin
+    pos_counts_idx = np.where(hist_counts > 0)[0]
+    last_binedge = pos_counts_idx[-1]
+    p2 = [bin_centers[last_binedge], hist_counts[last_binedge]]
+
+    best_threshold = -np.inf
+    max_dist = -np.inf
+    for idx in range(max_idx, last_binedge, 1):
+        x0 = bin_centers[idx]
+        y0 = hist_counts[idx]
+        a = [p1[0] - p2[0], p1[1] - p2[1]]
+        b = [x0 - p2[0], y0 - p2[1]]
+        cross_ab = a[0] * b[1] - b[0] * a[1]
+        per_dist = np.linalg.norm(cross_ab) / np.linalg.norm(a)
+        if per_dist > max_dist:
+            best_threshold = x0
+            max_dist = per_dist
+    assert best_threshold > -np.inf, 'Error in unimodal thresholding'
+    return best_threshold
+
+
+def unimodal_thresholding(input_image, str_elem_size=3):
```

Comment:
To have matching naming convention for each mask option maybe this could be something like create_unimodal_mask?